base/compiler/abstractlattice.jl

```diff
@@ -0,0 +1,127 @@
+abstract type AbstractLattice; end
+function widen end
+
+"""
+    struct JLTypeLattice
+
+A singleton type representing the lattice of Julia types, without any inference
+extensions.
+"""
+struct JLTypeLattice <: AbstractLattice; end
+widen(::JLTypeLattice) = error("Type lattice is the least-precise lattice available")
+
+"""
+    struct ConstsLattice
+
+A lattice extending `JLTypeLattice` and adjoining `Const` and `PartialTypeVar`.
+"""
+struct ConstsLattice <: AbstractLattice; end
+widen(::ConstsLattice) = JLTypeLattice()
+
+"""
+    struct PartialsLattice{L}
+
+A lattice extending lattice `L` and adjoining `PartialStruct` and `PartialOpaque`.
+"""
+struct PartialsLattice{L <: AbstractLattice} <: AbstractLattice
+    parent::L
+end
+widen(L::PartialsLattice) = L.parent
+
+"""
+    struct PartialsLattice{L}
+
+A lattice extending lattice `L` and adjoining `Conditional`.
+"""
+struct ConditionalsLattice{L <: AbstractLattice} <: AbstractLattice
+    parent::L
+end
+widen(L::ConditionalsLattice) = L.parent
+
+"""
+    struct OptimizerLattice
+
+The lattice used by the optimizer. Equivalent to
+`ConditionalsLattice{PartialsLattice{ConstsLattice}}`, but wrapped in a struct.
```

Comment:
Technically `Conditional` should never appear during optimization but `InterConditional` can appear as a return type. I'm not sure we want to have additional lattice layer for separating them though.

Reply:
I was thinking about this. Let me work through this some more and see which way looks nicer.